### PR TITLE
Feature: `<a>` tags in event description

### LINKED ISSUE #4085: Events content overflowing in Safari
# Problem
Events info is overflowing into the next event in the next event doesn't have an image in Safari (Version 14.1.2). It's fine in Chrome.

https://tippie.prod.drupal.uiowa.edu/events

Safari:
image
image

Chrome:
image

# Known Issue
https://sitenow.uiowa.edu/known-issues#uiowa-issue-4085

### LINKED ISSUE #4477: As a site owner, I can display an events list with teaser text that allows `<a>` tags
# Story

As a site owner, I can display an events list from events.uiowa.edu with teaser text that allows `<a>` tags so that site visitors can click on register and zoom links without clicking through to the next page. 

# Proposed solution

- Update docroot/modules/custom/sitenow_events/includes/sitenow_events.teaser.inc to allow `<a>` tags by updating `Xss::filter` to include an allowed list or switch over to use `Xss::filterAdmin` which is what the full display uses for better or worse.  
- For reference https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Component%21Utility%21Xss.php/class/Xss/9.2.x

### PR DESCRIPTION
**Pull request by bspeare**
Resolves https://github.com/uiowa/uiowa/issues/4477. 
Resolves https://github.com/uiowa/uiowa/issues/4085. 

# How to test

This pr changes the events module to use the `description` instead of the `description_text` from content hub.

- Test on any site using an events block.   
- You should see the text formatted and links active if they haven't been truncated.   
- Libraries has a lot of links if you set the event count to 20 for testing.

### CODE REVIEW

**Comment by cory-skeers** [comment]
Pulling down to test

**Review by cory-skeers — APPROVED** [review]
Tested and everything is working well! (Thanks for the extra tip of using libraries as a test case--very easy to find some links there.)

Both the enhancement and bug listed have been resolved. 💯

**Comment by bspeare** [comment]
@cory-skeers thanks for testing!  I made one last minute update to only allow certain tags.   Do you mind taking another glance and let me know if you see any issues with those code changes?

**Comment by joewhitsitt** [comment]
Another thing to consider is that the HTML tags count towards the 500 character truncate and may potentially cause broken HTML...

**Comment by cory-skeers** [comment]
@bspeare Looking at the extra change, I think everything looks good! I can pull it down again to make sure, but I'll be a little bit.

**Comment by cory-skeers** [comment]
> Another thing to consider is that the HTML tags count towards the 500 character truncate and may potentially cause broken HTML...

Ooh...also a good point to think about.

**Comment by bspeare** [comment]
Moving this back to draft to see if we can figure out a solution that uses `smart_trim`.

**Comment by joewhitsitt** [comment]
Yay! I think this smart_trim use will work nicely. I added the following description to test with after the xss filter.

```php
$variables["event"]["description"] = 'Nulla neque dolor, sagittis eget, iaculis quis, molestie non, velit. Donec vitae sapien ut libero venenatis faucibus. Fusce ac felis sit amet ligula pharetra condimentum. Curabitur vestibulum aliquam leo. Vestibulum suscipit nulla quis orci. Cras risus ipsum, faucibus ut, ullamcorper id, varius ac, leo. Nunc nulla. Nullam sagittis. Praesent vestibulum dapibus nibh. In <strong><a href="https://google.com">consectetuer turpis ut velit. This is my longer link that starts before the truncate cutoff</a></strong>';
```

And this is what I got. Nicely corrected HTML and it doesn't seem to count markup against the character count.

Screen Shot 2021-12-01 at 3 24 25 PM

File: docroot/modules/custom/sitenow_events/includes/sitenow_events.teaser.inc [hunk 1]
```diff
@@ -81,8 +81,14 @@ function template_preprocess_sitenow_events_teaser(array &$variables) {
...
-    $variables['event']['description_text'] = Unicode::truncate($variables['event']['description_text'], 500, TRUE, TRUE);
+  if (isset($variables['event']['description'])) {
+    // Limit what html is allowed.
+    $allowed_tags = ['a', 'strong', 'em'];
+    $variables['event']['description'] = Xss::filter($variables['event']['description'], $allowed_tags);
+
+    // Use smart_trim's helper to safely truncate without breaking HTML.
+    $truncate = new TruncateHTML();
+    $trim_suffix = '...';
+    $variables['event']['description'] = $truncate->truncateChars($variables['event']['description'], 400, $trim_suffix);
```
Comment (cory-skeers) [inline docroot/modules/custom/sitenow_events/includes/sitenow_events.teaser.inc]: Was the change from 500 characters to 400 intentional?

Comment (bspeare) [reply docroot/modules/custom/sitenow_events/includes/sitenow_events.teaser.inc]: Good point! I think I forgot to commit that change back.

Comment (bspeare) [reply docroot/modules/custom/sitenow_events/includes/sitenow_events.teaser.inc]: I changed it back to 500, are we good to merge this in?

Comment (cory-skeers) [reply docroot/modules/custom/sitenow_events/includes/sitenow_events.teaser.inc]: I think so!

**Pull request merged**